### DISCUSSION TITLE
WCF service with session over WsHttp

### DISCUSSION THREAD

**Discussion by AndersonGimenez86** [CoreWCF/CoreWCF#1480] [Q&A]
Hi,

I am wondering if is possible to use Sessions in CoreWcf with WsHttp bindings?
I guess that it´s not the same of ReliableSession which it´s not available in CoreWcf.

For instance what I need:
![image](https://github.com/user-attachments/assets/6b866ed9-8578-42ce-9d74-5fdad179936f)

I have noticed that it works if I only set SessionMode property on the server service contract, on the other hand when I set the property on client service contract it throws the following exception:

![image](https://github.com/user-attachments/assets/03ad74f8-93fc-4e71-9e10-b47562ae5747)

I found an issue opened for the same thing that I need but for NetTcp binding: [WCF service with session over NetTcp](https://github.com/CoreWCF/CoreWCF/issues/1124)

Also, an example of how to implement with NetTcp binding: [
CoreWCF-NetTcpSession-Sample](https://github.com/MontyGvMC/CoreWCF-NetTcpSession-Sample)

Could you please help me to understand if it is possible and how to do?

If no, what would be the option?

Many thanks.

**Comment by mconnew** [comment]
Yes, you can use sessions with WSHttpBinding. It uses WS-SecureConversation to establish the session. The easiest way is to use security mode TransportWithMessageCredentials and configure your credential type on WSHttpBinding.Security.Message. If you are free to pick your binding, you might want to consider NetHttpBinding and configure the websocket settings to always use them. That will establish a persistent connection using a websocket so will give you a transport session.  

I'm sorry I missed your prior issue, I see it's from last year. Ping me with a mention next week and I should have a bit of time to look at your repo.

**Reply by AndersonGimenez86** [reply]
Thanks for your answer and helping me.
About the prior issue, it´s Ok, we found a solution.

After changing the binging like you referred, the previous error has disappear.
The service builds and runs, but fails to map the endpoints:

![image](https://github.com/user-attachments/assets/86916fe6-4920-4361-ac99-060bd418eaf8)

Therefore, the operations are not available:

![image](https://github.com/user-attachments/assets/5df3735f-ff20-412d-bb43-797e1b6f8c07)

Current binding:
![image](https://github.com/user-attachments/assets/6dfa481d-792e-430e-bd0e-6eedbfe10d72)

About  NetHttpBinding and configure the websocket settings, do you have an example or link?
Sorry, I am struggling to convert to NetHttpBinding.

Thanks

**Reply by mconnew** [reply]
You just set `NetHttpBinding.WebSocketSettings.TransportUsage = WebSocketTransportUsage.Always;` on client and server. The error you're getting about unable to map the branch app is odd. It's checking IServiceDispatcher.SupportedChannelTypes against a bunch of different shapes supported for HTTP. It's supports pretty much all of them, so I don' know what shape is missing.

**Comment by fbenoit-coexya** [comment]
So there's no way to have sessions on WSHttpBinding without security like legacy WCF could?

**Reply by mconnew** [reply]
You should be able to do it the same way you did with WCF. We support WS-SecureConversation which can be used to establish a session for a sessionless transport such as HTTP. You do need to use HTTPS so need a service side certificate, but that's the same as for WCF. There is a risk though as it establishes a session on the service which is cached in a session cache. Having no client side authentication means any client, including malicious clients, can cause memory allocation on your server. If the server isn't running in a trusted network, ie exposed on the internet, it becomes a trivial DoS attack as they can either consume your max sessions quota, or if there's no session limit, consume an unbounded amount of RAM by creating as many sessions as possible until you run out of memory, or are swapping to disk enough to kill performance.

**Comment by bellamarshall** [comment]
Hi,

I'm migrating a WCF service that uses `SessionMode = SessionMode.Required` on the ServiceContract. It uses `WSHttpBinding` as defined below:

```
var wsHttpBinding = new WSHttpBinding
{
   MaxReceivedMessageSize = Properties.Settings.Default.MaxReceivedMessageSize,
   MaxBufferPoolSize = Properties.Settings.Default.MaxBufferPoolSize,
};
wsHttpBinding.Security.Mode = SecurityMode.None;
wsHttpBinding.ReliableSession.Enabled = true;
wsHttpBinding.ReceiveTimeout = TimeSpan.FromSeconds(Properties.Settings.Default.SessionTimeoutSeconds);
wsHttpBinding.ReliableSession.InactivityTimeout = TimeSpan.FromSeconds(Properties.Settings.Default.SessionTimeoutSeconds);
```

I'm attempting to use an equivalent CoreWCF binding (minus ReliableSessions) to register my endpoint like so:

```
serviceBuilder.AddService<Service>(serviceOptions =>
{
   serviceOptions.DebugBehavior.IncludeExceptionDetailInFaults = true;
   serviceOptions.BaseAddresses.Add(new Uri($"http://localhost:{port}"));
});
serviceBuilder.AddServiceEndpoint<Service, IService>(wsHttpBinding, "/Service");
```

but the service is failing to map the end point

<img width="824" height="100" alt="image" src="https://github.com/user-attachments/assets/25a5e0d1-ceff-4545-9665-c5ccf02fe529" />

Is there a step I'm missing or is this not supported?

Thanks so much
